content/ui-ux.md

```diff
@@ -0,0 +1,514 @@
+
+A final place where you may want to translate is the actual data in your database. Of course how you might go about this would be very much a problem unique to your application!
+
+### Event handling
+
+A large part of your UI involves responding to events initated by the user, and there are some steps that you should take to make sure your application does this well
+
+#### Throttling method calls on user action
+
+It's typical to make some kind of change to the database when a user takes an action. However it's important to make sure you don't do this *too* much. For instance, if you wish to save the user's text as they type in a textbox, you should take steps to make sure that you don't send method calls to your database more than every few hundred milliseconds.
+
+If you do not, you'll see performance problems across the board; you'll be flooding the user's network connection with a lot of minute changes, the UI will update on every keystroke, potentially causing poor performance, and your database will suffer with a lot of writes.
+
+To throttle writes, a typical approach is to use underscore's [`.throttle()`](http://underscorejs.org/#throttle) or [`.debounce()`](http://underscorejs.org/#debounce) functions. For instance, in the Todos example app, we throttle writes on user input to 300ms:
+
+```js
+Template.todosItem.helpers({
+  // update the text of the item on keypress but throttle the event to ensure
+  // we don't flood the server with updates (handles the event at most once
+  // every 300ms)
+  'keyup input[type=text]': _.throttle(function(event) {
+    Todos.methods.updateText.call({
+      todoId: this.todo._id,
+      newText: event.target.value
+    }, (err) => {
+      err && alert(err.error);
+    });
+  }, 300)
+});
+```
+
+#### Limiting re-rendering
+
+Even if you aren't saving data over the wire to the database on every user input, sometimes you still may wish to update in-memory data stores on every user change. In theory this is fine, but in practice sometimes if updating that data store triggers a lot of UI changes, you can see poor performance (and missed keystrokes) when you do that. In such cases you can limit re-rendering by throttling such changes in a similar way to the way we throttle the method call above.
+
+#### Scroll events
+
+Another type of event that fires very frequently is a scroll event. Such event trigger as a user scrolls up and down a page and are commonly used for "waypoint" behaviour. For example in the Localmarket example app, we use the [jquery-waypoints](https://atmospherejs.com/meteor/jquery-waypoints) library to trigger a class being attached to the nav element when you scroll past a certain point:
```

Comment:
Yeah, I guess I thought we hand coded it in the sample app but we don't there. We do some hand coding but not in any open source apps...